src/runtime/ndarray.cc

```diff
@@ -206,6 +213,46 @@ void NDArray::CopyFromTo(DGLArray* from, DGLArray* to) {
       from->dtype);
 }
 
+void NDArray::RecordedCopyFromTo(DGLArray* from, DGLArray* to, void* pyt_ctx) {
+  size_t from_size = GetDataSize(*from);
+  size_t to_size = GetDataSize(*to);
+  CHECK_EQ(from_size, to_size)
+      << "DGLArrayCopyFromTo: The size must exactly match.";
+
+  CHECK(
+      from->ctx.device_type == to->ctx.device_type ||
+      from->ctx.device_type == kDGLCPU || to->ctx.device_type == kDGLCPU)
+      << "Can not copy across different ctx types directly.";
```

Comment:
Why can't I copy from one GPU to another? Also, the error message here only captures the first line of the check.

Reply:
This checker doesn't block GPU-GPU copy. GPU-GPU copy is the copy across the same ctx type.
It blocks the copy across 2 non-host device types, which I guess is one of the heuristics in the code. In most runs, other than cpu, we only have one device (CUDA). Thus, maybe we can simplify the logic here...